module/zfs/vdev_removal.c

```diff
@@ -1995,17 +1995,18 @@ spa_vdev_remove_top_check(vdev_t *vd)
 		if (available < vd->vdev_stat.vs_alloc)
 			return (SET_ERROR(ENOSPC));
 	} else {
-		/* available space in the pool's normal class */
+		/*
+		 * This is a normal device. There has to be enough free
+		 * space in the pool's normal class to remove the device
+		 * and leave 1.6% of the space available.  The available
+		 * space accounts for both reserved and "slop" space.
+		 */
 		uint64_t available = dsl_dir_space_available(
 		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE);
-		if (available <
-		    vd->vdev_stat.vs_dspace + spa_get_slop_space(spa)) {
-			/*
-			 * This is a normal device. There has to be enough free
-			 * space to remove the device and leave double the
-			 * "slop" space (i.e. we must leave at least 3% of the
-			 * pool free, in addition to the normal slop space).
-			 */
+
+		if (available < vd->vdev_stat.vs_dspace ||
+		    available - vd->vdev_stat.vs_dspace <
+		    spa_get_dspace(spa) >> 6) {
```

Comment:
I don't think this fixes the problem that removing a disproportionately-huge device can get ENOSPC when we want it to succeed.  e.g. imagine a pool with a 1TB device and a 100TB device, with 0.5TB used.  The check would be 100.5 - 100 < 101*0.016, or 0.5 < 1.6, which is true.  I think want to require that *after removal*, there is 1.6% space AVAIL.

Reply:
While you're right, @ahrens, and I'm on an old 2.0.4 version and currently unwilling to upgrade, if nothing has changed in this regards since 2.0.4 then a major tighten up would be massively beneficial, and/or a way to override this with -f when the user does indeed know better. I will tag you both where I record my story from the last 2 days.

Reply:
https://github.com/openzfs/zfs/issues/13552#issuecomment-1413386598